### PR TITLE
Add homework 02 task 04 and test 04

### PR DESCRIPTION
**Pull request by mukhaw**

### CODE REVIEW

File: hm_1/tasks/task_02_04.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+cache_func = cache(func)
+some = 100, 200
+val_1 = cache_func(*some)
+val_2 = cache_func(*some)
+assert val_1 is val_2
+"""
+from typing import Callable
+
+
+def func(a, b):
```
Comment (Cjkjvfnby) [inline hm_1/tasks/task_02_04.py:16]: You can remove `func` and start using mocks.

**Review by Cjkjvfnby — CHANGES_REQUESTED** [review]
Please add test with argument: `[1, 2]`

File: hm_1/tasks/task_02_04.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        if args in caching:
+            return caching[args]
+        mock.metod.return_value = (args[0] ** args[1]) ** 2
+        caching[args] = mock.method()
+        return caching[args]
+
+    return cache_func
+
+
+mock = Mock()
```
Comment (Cjkjvfnby) [inline hm_1/tasks/task_02_04.py:30]: Please don't use that mock here.  It is supposed to be used only in test_* files.

File: hm_1/tasks/task_02_04.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+some = 100, 200
+val_1 = cache_func(*some)
+val_2 = cache_func(*some)
+assert val_1 is val_2
+"""
+from typing import Callable
+from unittest.mock import Mock
+
+
+def cache(func: Callable) -> Callable:
```
Comment (Cjkjvfnby) [inline hm_1/tasks/task_02_04.py:17]: Do you use argument `func` anywhere?

**Review by Cjkjvfnby — CHANGES_REQUESTED** [review]

File: hm_1/tasks/task_02_04.py (hunk 3 above)
Comment (mukhaw) [reply hm_1/tasks/task_02_04.py:17]: no, I can remove it right, because I use mock?

File: hm_1/tasks/task_02_04.py (hunk 2 above)
Comment (mukhaw) [reply hm_1/tasks/task_02_04.py:30]: I understand right, that I must include mock in the test but in the task's code, I don't change anything?

Comment (Cjkjvfnby) [reply hm_1/tasks/task_02_04.py:30]: Yes.

File: test/test_02_04.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
+from unittest.mock import Mock
+
+from hm_1.tasks.task_02_04 import cache
+
+
+def func(a, b):
+    return (a ** b) ** 2
+
+
+mock = Mock(side_effect=func)
```
Comment (Cjkjvfnby) [inline test/test_02_04.py:10]: Move this line inside each test. And call mock.assert_called_once to ensure a single call.

Mock is mutated during calls, you do not want all tests to depend on a single mutable object, each test should have its own mock.

File: test/test_02_04.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    cache_func = cache(mock)
+    some = 0, 20
+    val_1 = cache_func(*some)
+    val_2 = cache_func(*some)
+    assert val_1 is val_2
+
+
+def test_cache_return_cache_func_cached_value_0_0():
+    cache_func = cache(mock)
+    mock.some = 0, 0
```
Comment (Cjkjvfnby) [inline test/test_02_04.py:32]: This is the same type of test as others.  It does not bring additional value.

Tests with different pair of arguments probably do not affect any caching features.

File: hm_1/tasks/task_02_04.py [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    caching = {}
+
+    def cache_func(*args):
+        if args in caching:
+            return caching[args]
+        caching[args] = func(*args)
+        return caching[args]
```
Comment (Cjkjvfnby) [inline hm_1/tasks/task_02_04.py:23]: If you rearrange the order of operations you can make the code a little bit simpler.

**Review by Cjkjvfnby — CHANGES_REQUESTED** [review]

**Pull request merged**